File: ruby/hash_map.rb
require_relative 'linked_list'

class HashMap
  def initialize
    @store = Array.new(8) {LinkedList.new}
    @count = 0
  end

  def add(key, val)
    delete(key) #don't add duplicate keys
    resize! if num_buckets == @count
    bucket(key).insert(key, val)
    @count += 1
    key
  end

  def delete(key)
    removed = bucket(key).remove(key)
    @count -= 1 if removed
    removed
  end

  def get(key)
    bucket(key).find_node(key).val
  end

  def inspect
    res = []
    @store.each do |list|
      list.each do |node|
        res << "#{node.key} => #{node.val}"
      end
    end
    "{#{res.join(", ")}}"
  end

  private
  def num_buckets
    @store.length
  end

  def bucket(key)
    @store[key.hash % num_buckets]
  end

  def resize!
    old_store = @store
    @count = 0
    @store = Array.new(num_buckets * 2) {LinkedList.new}
    old_store.each do |list|
      list.each do |node|
        self.add(node.key, node.val)
      end
    end
  end
end

File: javascript/hash_map.js
/* global LinkedList */

class HashMap {
  constructor() {
    this.store = [];
    for (var i = 0; i < 8; i++) {this.store.push(new LinkedList);}

    this.count = 0;
  }

  set(key, val) {
    this.delete(key);
    if (this.count === this.numBuckets()) { this.resize(); }

    this.bucket(key).insert(key, val);
    this.count++;
    return key;
  }

  get(key) {
    this.bucket(key).get(key);
  }

  delete(key) {
    let val = this.bucket(key).delete(key);
    if (val !== null) {this.count--; }
    return val;
  }

  resize() {
    let oldStore = this.store;
    this.store = new Array(this.numBuckets *  2);
    this.count = 0;

    oldStore.forEach(bucket => {
      bucket.each(node => {
        this.set(node.key, node.val);
      });
    });
  }

  bucket(key) {
    // should hash key
    return this.store[key % this.numBuckets()];
  }

  numBuckets() {
    return this.store.length();
  }
}
